clients/src/main/java/org/apache/kafka/common/metrics/Sensor.java

```diff
@@ -291,10 +295,24 @@ public boolean hasExpired() {
     }
 
     /**
-     * KafkaMetrics of sensors which use SampledStat should be synchronized on the Sensor object
-     * to allow concurrent reads and updates. For simplicity, all sensors are synchronized on Sensor.
+     * KafkaMetrics of sensors which use SampledStat should be synchronized on the same lock
+     * for sensor record and metric value read to allow concurrent reads and updates. For simplicity,
+     * all sensors are synchronized on this object.
+     * <p>
+     * Sensor object is not used as a lock since metrics reporter is invoked while holding Sensor
+     * and Metrics locks to report addition and removal of metrics and synchronized reporters may
+     * deadlock if Sensor lock is used for reading metrics values.
+     * </p><p>
+     * Locking order (assume all MetricsReporter methods may be synchronized):
+     * <ul>
+     *   <li>Sensor#add: Sensor -> Metrics -> MetricsReporter</li>
```

Comment:
It would be useful to document that the Sensor object itself is used as a lock to protect the access to stats and metrics.

Reply:
@junrao Thanks for the review, I have added that to the comment.